File: oolab/src/main/java/agh/ics/oop/model/Vector2D.java
package agh.ics.oop.model;

public record Vector2D(int x, int y) {

    public String toString() {
        return "(" + x + "," + y + ")";
    }

    public boolean precedes(Vector2D other) {
        return other.x >= x() && other.y >= y();
    }

    public boolean follows(Vector2D other) {
        return other.x <= x() && other.y <= y();
    }

    public Vector2D add(Vector2D other) {
        return new Vector2D(x() + other.x, y() + other.y);
    }

    public Vector2D subtract(Vector2D other) {
        return new Vector2D(x() - other.x, y() - other.y);
    }

    public Vector2D upperRight(Vector2D other) {
        return new Vector2D(Math.max(x(), other.x), Math.max(y(), other.y));
    }

    public Vector2D lowerLeft(Vector2D other) {
        return new Vector2D(Math.min(x(), other.x), Math.min(y(), other.y));
    }

    public Vector2D opposite() {

        return new Vector2D((-1) * x(), (-1) * y());
    }


    @Override
    public boolean equals(Object other) {
        if (this == other)
            return true;
        if (!(other instanceof Vector2D))
            return false;
        Vector2D that = (Vector2D) other;
        return x == that.x && y == that.y;
    }


}

File: kotlin_lab/untitled/src/main/kotlin/model/kotlin/Vector2d.kt
package model.kotlin

import kotlin.math.max
import kotlin.math.min

fun MapDirection.toUnitVector(): Vector2d = when (this) {
    MapDirection.NORTH -> Vector2d(0, 1);
    MapDirection.SOUTH -> Vector2d(0, -1);
    MapDirection.WEST -> Vector2d(-1, 0);
    MapDirection.EAST -> Vector2d(1, 0);
}

data class Vector2d(val x: Int, val y: Int) : Comparable<Vector2d> {
    fun opposite() = Vector2d((-1) * x, (-1) * y)

    operator fun plus(addVector: Vector2d) = Vector2d(addVector.x + x, addVector.y + y)

    operator fun minus(subVector: Vector2d) = Vector2d(x - subVector.x, y - subVector.y)

    fun upperRight(other: Vector2d) = Vector2d(max(x, other.x), max(y, other.y))

    fun lowerLeft(other: Vector2d) = Vector2d(min(x, other.x), min(y, other.y))

    override fun compareTo(other: Vector2d): Int = when {
        x >= other.x && y >= other.y -> 1
        x <= other.x && y <= other.y -> -1
        else -> 0
    }


}
